File: sdk/core/llm/formatters/anthropic.ts
/**
 * Anthropic Format Converter
 *
 * Implements the conversion of request and response formats for the Anthropic API
 */

import { BaseFormatter } from "./base.js";
import type { LLMRequest, LLMResult, LLMMessage, LLMToolCall, ToolCallFormat } from "@wf-agent/types";
import type { ToolSchema } from "@wf-agent/types";
import type { FormatterConfig, BuildRequestResult, ParseStreamChunkResult } from "./types.js";
import { convertToolsToAnthropicFormat } from "../utils/index.js";
import { ToolDeclarationFormatter } from "@wf-agent/prompt-templates";
import { getToolCallParserOptions } from "./tool-format-selector.js";
import { ToolCallParser } from "./tool-call-parser.js";
import { HistoryConverter } from "../../messages/history-converter.js";

/**
 * Anthropic format converter
 */
export class AnthropicFormatter extends BaseFormatter {
  private readonly apiVersion: string;

  constructor(apiVersion: string = "2023-06-01") {
    super();
    this.apiVersion = apiVersion;
  }

  getSupportedProvider(): string {
    return "ANTHROPIC";
  }

  /**
   * Build request in native function-calling mode
   */
  protected buildNativeRequest(request: LLMRequest, config: FormatterConfig): BuildRequestResult {
    const body = this.buildRequestBody(request, config);

    // Constructing request headers
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "anthropic-version": (config.profile.metadata?.["apiVersion"] as string) || this.apiVersion,
      "anthropic-dangerous-direct-browser-access": "false",
      ...config.profile.headers,
    };

    // Add authentication headers (supporting both x-api-key and Authorization Bearer methods)
    if (config.profile.apiKey) {
      Object.assign(headers, this.buildAuthHeader(config.profile.apiKey, config, "x-api-key"));
    }

    // Add custom request headers
    Object.assign(headers, this.buildCustomHeaders(config));

    // Apply a custom request body
    const finalBody = this.applyCustomBody(body, config);

    // Construct query parameters
    const queryString = this.buildQueryString(config);

    return {
      httpRequest: {
        url: `/v1/messages${queryString}`,
        method: "POST",
        headers,
        body: finalBody,
        timeout: config.timeout,
      },
      transformedBody: finalBody,
    };
  }

  /**
   * Build request in text-based mode (XML/JSON)
   */
  protected override buildTextModeRequest(request: LLMRequest, config: FormatterConfig): BuildRequestResult {
    const format = this.getToolCallFormat(config);

    // 1. Generate tool declarations
    const toolDeclarations = ToolDeclarationFormatter.formatTools(
      request.tools || [],
      {
        format: format === 'json_wrapped' || format === 'json_raw' ? 'json' : 'xml',
        xmlTags: config.toolCallFormat?.xmlTags,
        markers: config.toolCallFormat?.markers,
        includeDescription: config.toolCallFormat?.includeDescription,
      }
    );

    // 2. Convert history to text mode (if needed)
    const convertedMessages = HistoryConverter.convertToTextMode(
      request.messages,
      format,
      {
        xmlTags: config.toolCallFormat?.xmlTags,
        markers: config.toolCallFormat?.markers,
      }
    );

    // 3. For Anthropic, inject tools into system field
    const { systemMessage, filteredMessages } = this.extractSystemMessage(convertedMessages);
    const existingSystem = systemMessage?.content || '';
    const instructions = this.getToolUsageInstructions(format);
    
    const systemContent = `${existingSystem}\n\n${instructions}\n\n${toolDeclarations}`.trim();

    // 3. Build request WITHOUT tools field
    const parameters = this.mergeParameters(config.profile.parameters, request.parameters);
    const body: Record<string, unknown> = {
      model: config.profile.model,
      messages: this.convertMessages(filteredMessages),
      system: systemContent,
      stream: config.stream || false,
    };

    const { stream: _stream, ...otherParams } = parameters;
    Object.assign(body, otherParams);

    // NO tools field!

    // Constructing request headers
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "anthropic-version": (config.profile.metadata?.["apiVersion"] as string) || this.apiVersion,
      "anthropic-dangerous-direct-browser-access": "false",
      ...config.profile.headers,
    };

    // Add authentication headers
    if (config.profile.apiKey) {
      Object.assign(headers, this.buildAuthHeader(config.profile.apiKey, config, "x-api-key"));
    }

    // Add custom request headers
    Object.assign(headers, this.buildCustomHeaders(config));

    // Apply a custom request body
    const finalBody = this.applyCustomBody(body, config);

    // Construct query parameters
    const queryString = this.buildQueryString(config);

    return {
      httpRequest: {
        url: `/v1/messages${queryString}`,
        method: "POST",
        headers,
        body: finalBody,
        timeout: config.timeout,
      },
      transformedBody: finalBody,
    };
  }

  /**
   * Parse response in native function-calling mode
   */
  protected parseNativeResponse(data: unknown, config: FormatterConfig): LLMResult {
    const dataRecord = data as Record<string, unknown>;
    const content = this.extractContent(dataRecord["content"]);
    const toolCalls = this.extractToolCalls(dataRecord["content"]);

    // Extract thinking content (for Claude extended thinking)
    const thinkingContent = this.extractThinkingContent(dataRecord["content"]);

    const usage = dataRecord["usage"] as Record<string, number> | undefined;

    return {
      id: dataRecord["id"] as string,
      model: dataRecord["model"] as string,
      content,
      message: {
        role: "assistant",
        content,
        toolCalls,
      },
      toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      usage: usage
        ? {
            promptTokens: usage["input_tokens"] ?? 0,
            completionTokens: usage["output_tokens"] ?? 0,
            totalTokens: (usage["input_tokens"] ?? 0) + (usage["output_tokens"] ?? 0),
          }
        : undefined,
      finishReason: dataRecord["stop_reason"] as string,
      duration: 0,
      metadata: {
        type: dataRecord["type"] as string,
        stopReason: dataRecord["stop_reason"] as string,
      },
      reasoningContent: thinkingContent,
    };
  }

  /**
   * Parse response in text-based mode (XML/JSON)
   */
  protected override parseTextModeResponse(data: unknown, config: FormatterConfig): LLMResult {
    const dataRecord = data as Record<string, unknown>;
    const content = this.extractContent(dataRecord["content"]);
    const format = this.getToolCallFormat(config);

    // Parse tool calls from content using ToolCallParser
    const toolCalls = config.toolCallFormat
      ? ToolCallParser.parseFromText(
          content,
          getToolCallParserOptions(config.toolCallFormat.format, config.toolCallFormat.markers)
        )
      : [];

    // Extract thinking content (for Claude extended thinking)
    const thinkingContent = this.extractThinkingContent(dataRecord["content"]);

    const usage = dataRecord["usage"] as Record<string, number> | undefined;

    return {
      id: dataRecord["id"] as string,
      model: dataRecord["model"] as string,
      content,
      message: {
        role: "assistant",
        content,
        toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      },
      toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      usage: usage
        ? {
            promptTokens: usage["input_tokens"] ?? 0,
            completionTokens: usage["output_tokens"] ?? 0,
            totalTokens: (usage["input_tokens"] ?? 0) + (usage["output_tokens"] ?? 0),
          }
        : undefined,
      finishReason: dataRecord["stop_reason"] as string,
      duration: 0,
      metadata: {
        type: dataRecord["type"] as string,
        stopReason: dataRecord["stop_reason"] as string,
      },
      reasoningContent: thinkingContent,
    };
  }

  /**
   * Get tool usage instructions based on format
   */
  private getToolUsageInstructions(format: ToolCallFormat): string {
    if (format === 'xml') {
      return `## Tool Usage Instructions

When you need to use a tool, format your response as follows:

<tool_use>
  <tool_name>tool_name_here</tool_name>
  <parameters>
    <param1>value1</param1>
    <param2>value2</param2>
  </parameters>
</tool_use>

You can use multiple tools in one response by including multiple <tool_use> blocks.`;
    } else if (format === 'json_wrapped') {
      return `## Tool Usage Instructions

When you need to use a tool, format your response as follows:

<<<TOOL_CALL>>>
{
  "tool": "tool_name_here",
  "parameters": {
    "param1": "value1",
    "param2": "value2"
  }
}
<<<END_TOOL_CALL>>>

You can use multiple tools in one response by including multiple blocks.`;
    }

    return '';
  }

  parseStreamChunk(data: unknown): ParseStreamChunkResult {
    const dataRecord = data as Record<string, unknown>;
    const delta = dataRecord["delta"] as Record<string, unknown> | undefined;
    const contentBlock = dataRecord["content_block"] as Record<string, unknown> | undefined;
    const usage = dataRecord["usage"] as Record<string, number> | undefined;
    const message = dataRecord["message"] as Record<string, unknown> | undefined;
    const messageUsage = message?.["usage"] as Record<string, number> | undefined;

    switch (dataRecord["type"]) {
      case "content_block_delta":
        // Text delta event
        if (delta && delta["type"] === "text_delta" && delta["text"]) {
          return {
            chunk: {
              delta: delta["text"] as string,
              done: false,
              raw: data,
            },
            valid: true,
          };
        }
        // Thinking delta event (for Claude extended thinking)
        if (delta && delta["type"] === "thinking_delta" && delta["thinking"]) {
          return {
            chunk: {
              delta: "",
              done: false,
              reasoningDelta: delta["thinking"] as string,
              raw: data,
            },
            valid: true,
          };
        }
        break;

      case "content_block_start":
        // Content block start event (tool call)
        if (contentBlock && contentBlock["type"] === "tool_use") {
          const toolCall: LLMToolCall = {
            id: contentBlock["id"] as string,
            type: "function",
            function: {
              name: contentBlock["name"] as string,
              arguments: JSON.stringify(contentBlock["input"] || {}),
            },
          };
          return {
            chunk: {
              delta: "",
              done: false,
              toolCallsDelta: [toolCall],
              raw: { toolCall },
            },
            valid: true,
          };
        }
        // Thinking block start event
        if (contentBlock && contentBlock["type"] === "thinking") {
          return {
            chunk: {
              delta: "",
              done: false,
              raw: data,
            },
            valid: true,
          };
        }
        break;

      case "message_delta":
        // Message delta event (contains usage)
        if (usage) {
          // Handle stop_reason: end_turn, max_tokens, stop_sequence, tool_use
          const stopReason = delta?.["stop_reason"] as string | undefined;
          const isDone =
            stopReason === "end_turn" ||
            stopReason === "max_tokens" ||
            stopReason === "stop_sequence" ||
            stopReason === "tool_use";
          return {
            chunk: {
              delta: "",
              done: isDone,
              usage: {
                promptTokens: usage["input_tokens"] || 0,
                completionTokens: usage["output_tokens"] || 0,
                totalTokens: (usage["input_tokens"] || 0) + (usage["output_tokens"] || 0),
              },
              finishReason: stopReason,
              raw: data,
            },
            valid: true,
          };
        }
        break;

      case "message_start":
        // Message start event
        if (messageUsage) {
          return {
            chunk: {
              delta: "",
              done: false,
              usage: {
                promptTokens: messageUsage["input_tokens"] || 0,
                completionTokens: 0,
                totalTokens: messageUsage["input_tokens"] || 0,
              },
              raw: data,
            },
            valid: true,
          };
        }
        break;

      default:
        break;
    }

    return { chunk: { done: false }, valid: false };
  }

  convertTools(tools: ToolSchema[]): unknown {
    return convertToolsToAnthropicFormat(tools);
  }

  convertMessages(messages: LLMMessage[]): unknown[] {
    return messages
      .filter(msg => msg.role !== "system") // Anthropic does not support system messages in the messages array.
      .map(msg => {
        const converted: { role: string; content: unknown[] } = {
          role: msg.role === "assistant" ? "assistant" : "user",
          content: [],
        };

        // Process the content.
        if (typeof msg.content === "string") {
          converted.content.push({
            type: "text",
            text: msg.content,
          });
        } else if (Array.isArray(msg.content)) {
          converted.content = msg.content;
        }

        // Add the results of the tool calls.
        if (msg.role === "tool" && msg.toolCallId) {
          converted.role = "user";
          converted.content = [
            {
              type: "tool_result",
              tool_use_id: msg.toolCallId,
              content: typeof msg.content === "string" ? msg.content : JSON.stringify(msg.content),
            },
          ];
        }

        // Add tool calls
        if (msg.toolCalls && msg.toolCalls.length > 0) {
          converted.content = msg.toolCalls.map(call => ({
            type: "tool_use",
            id: call.id,
            name: call.function.name,
            input: JSON.parse(call.function.arguments),
          }));
        }

        return converted;
      });
  }

  parseToolCalls(toolCalls: unknown[]): LLMToolCall[] {
    if (!toolCalls) return [];
    return toolCalls.map(call => ({
      id: (call as { id: string }).id,
      type: "function",
      function: {
        name: (call as { name: string }).name,
        arguments:
          typeof (call as { input?: unknown }).input === "string"
            ? (call as { input: string }).input
            : JSON.stringify((call as { input?: unknown }).input || {}),
      },
    }));
  }

  /**
   * Construct a Token count request
   * @param request: LLM request
   * @param config: Format converter configuration
   * @returns: HTTP request configuration
   */
  buildCountTokensRequest(request: LLMRequest, config: FormatterConfig): BuildRequestResult {
    const parameters = this.mergeParameters(config.profile.parameters, request.parameters);

    const body: Record<string, unknown> = {
      model: config.profile.model,
    };

    // Processing system messages
    const { systemMessage, filteredMessages } = this.extractSystemMessage(request.messages);
    if (systemMessage) {
      body["system"] =
        typeof systemMessage.content === "string"
          ? systemMessage.content
          : JSON.stringify(systemMessage.content);
    }
    body["messages"] = this.convertMessages(filteredMessages);

    // Add tools (if any).
    if (request.tools && request.tools.length > 0) {
      body["tools"] = this.convertTools(request.tools);
    }

    // Add the thinking configuration (if it exists).
    if (parameters["thinking"]) {
      body["thinking"] = parameters["thinking"];
    }

    // Add tool_choice (if it exists)
    if (parameters["tool_choice"]) {
      body["tool_choice"] = parameters["tool_choice"];
    }

    // Constructing request headers
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "anthropic-version": (config.profile.metadata?.["apiVersion"] as string) || this.apiVersion,
      "anthropic-dangerous-direct-browser-access": "false",
      ...config.profile.headers,
    };

    // Add authentication headers
    if (config.profile.apiKey) {
      Object.assign(headers, this.buildAuthHeader(config.profile.apiKey, config, "x-api-key"));
    }

    // Add custom request headers
    Object.assign(headers, this.buildCustomHeaders(config));

    return {
      httpRequest: {
        url: "/v1/messages/count_tokens",
        method: "POST",
        headers,
        body,
        timeout: config.timeout,
      },
      transformedBody: body,
    };
  }

  /**
   * Construct the request body
   */
  private buildRequestBody(request: LLMRequest, config: FormatterConfig): Record<string, unknown> {
    const parameters = this.mergeParameters(config.profile.parameters, request.parameters);

    const body: Record<string, unknown> = {
      model: config.profile.model,
      ["max_tokens"]: parameters["max_tokens"] || 4096,
      stream: config.stream || false,
    };

    // Handle system messages
    const { systemMessage, filteredMessages } = this.extractSystemMessage(request.messages);
    if (systemMessage) {
      body["system"] =
        typeof systemMessage.content === "string"
          ? systemMessage.content
          : JSON.stringify(systemMessage.content);
    }
    body["messages"] = this.convertMessages(filteredMessages);

    // Merge other parameters (excluding stream and max_tokens, which have been handled separately).
    const { stream: _stream, max_tokens: _max_tokens, ...otherParams } = parameters;
    Object.assign(body, otherParams);

    // Add tools
    if (request.tools && request.tools.length > 0) {
      body["tools"] = this.convertTools(request.tools);
    }

    // Handle thinking configuration (for Claude extended thinking)
    // Users can set parameters.thinking = { type: 'enabled', budget_tokens: 10000 }
    // or parameters.thinking = { type: 'adaptive', effort: 'high' }
    // The deepMerge will handle this correctly

    return body;
  }

  /**
   * Extract the text content.
   */
  private extractContent(content: unknown): string {
    if (!content || !Array.isArray(content)) {
      return "";
    }

    return content
      .filter(item => (item as { type?: string }).type === "text")
      .map(item => (item as { text?: string }).text ?? "")
      .join("");
  }

  /**
   * Extract tool calls
   */
  private extractToolCalls(content: unknown): LLMToolCall[] {
    if (!content || !Array.isArray(content)) {
      return [];
    }

    return content
      .filter(item => (item as { type?: string }).type === "tool_use")
      .map(item => ({
        id: (item as { id: string }).id,
        type: "function",
        function: {
          name: (item as { name: string }).name,
          arguments: JSON.stringify((item as { input?: unknown }).input || {}),
        },
      }));
  }

  /**
   * Extract thinking content blocks from Anthropic responses
   *
   *
   */
  private extractThinkingContent(content: unknown): string | undefined {
    if (!content || !Array.isArray(content)) {
      return undefined;
    }

    const thinkingBlocks = content.filter(item => (item as { type?: string }).type === "thinking");
    if (thinkingBlocks.length === 0) {
      return undefined;
    }

    return thinkingBlocks.map(block => (block as { thinking?: string }).thinking || "").join("\n");
  }
}

File: crates/wf-llm/src/formatters/anthropic.rs
use reqwest::Method;
use wf_types::llm::{LlmRequest, LlmResult as LlmResponseType, LlmProfile, MessageStreamEvent};
use crate::error::LlmResult;
use wf_types::tool::Tool;
use super::LlmFormatter;

pub struct AnthropicFormatter {
    base_url: String,
}

impl Default for AnthropicFormatter {
    fn default() -> Self {
        Self::new()
    }
}

impl AnthropicFormatter {
    pub fn new() -> Self {
        Self {
            base_url: "https://api.anthropic.com/v1".to_string(),
        }
    }

    pub fn with_base_url(base_url: String) -> Self {
        Self { base_url }
    }

    fn convert_messages(&self, messages: &[wf_types::message::Message]) -> Vec<serde_json::Value> {
        let mut result = Vec::new();
        let mut current_text: Option<String> = None;

        for msg in messages {
            let role = match msg.role {
                wf_types::message::MessageRole::System => continue,
                wf_types::message::MessageRole::User => "user",
                wf_types::message::MessageRole::Assistant => "assistant",
                wf_types::message::MessageRole::Tool => "user",
            };

            let content = match &msg.content {
                wf_types::message::MessageContentValue::Text(text) => {
                    if text.is_empty() { continue; }
                    serde_json::json!([{"type": "text", "text": text}])
                }
                wf_types::message::MessageContentValue::Rich(blocks) => {
                    let converted: Vec<serde_json::Value> = blocks.iter().map(|block| {
                        match block {
                            wf_types::message::MessageContent::Text { text } => {
                                serde_json::json!({"type": "text", "text": text})
                            }
                            wf_types::message::MessageContent::ToolUse { tool_use } => {
                                serde_json::json!({
                                    "type": "tool_use",
                                    "id": tool_use.id,
                                    "name": tool_use.name,
                                    "input": tool_use.input,
                                })
                            }
                            wf_types::message::MessageContent::ToolResult { tool_result } => {
                                serde_json::json!({
                                    "type": "tool_result",
                                    "tool_use_id": tool_result.tool_use_id,
                                    "content": tool_result.content,
                                    "is_error": tool_result.is_error.unwrap_or(false),
                                })
                            }
                            wf_types::message::MessageContent::ImageUrl { image_url } => {
                                let url = &image_url.url;
                                let (media_type, data) = if let Some(rest) = url.strip_prefix("data:") {
                                    if let Some(semicolon) = rest.find(';') {
                                        let mime = &rest[..semicolon];
                                        let b64 = rest[semicolon + 1..].strip_prefix("base64,").unwrap_or("").to_string();
                                        (mime.to_string(), b64)
                                    } else {
                                        ("image/png".to_string(), rest.to_string())
                                    }
                                } else {
                                    ("image/png".to_string(), url.clone())
                                };
                                serde_json::json!({
                                    "type": "image",
                                    "source": {
                                        "type": "base64",
                                        "media_type": media_type,
                                        "data": data,
                                    }
                                })
                            }
                            _ => serde_json::json!({"type": "text", "text": ""}),
                        }
                    }).collect();
                    serde_json::json!(converted)
                }
            };

            if role == "user" {
                if let Some(tool_id) = &msg.tool_call_id {
                    let entry = serde_json::json!({
                        "role": "user",
                        "content": [{
                            "type": "tool_result",
                            "tool_use_id": tool_id,
                            "content": match &msg.content {
                                wf_types::message::MessageContentValue::Text(t) => t.clone(),
                                _ => "".to_string(),
                            }
                        }]
                    });
                    result.push(entry);
                    continue;
                }
                if let Some(text) = &current_text {
                    let entry = serde_json::json!({
                        "role": "assistant",
                        "content": [{"type": "text", "text": text}]
                    });
                    result.push(entry);
                    current_text = None;
                }
            } else if role == "assistant" {
                if let Some(text) = &current_text {
                    let entry = serde_json::json!({
                        "role": "assistant",
                        "content": [{"type": "text", "text": text}]
                    });
                    result.push(entry);
                    current_text = None;
                }
                if let Some(ref tool_calls) = msg.tool_calls {
                    let mut blocks = Vec::new();
                    if let wf_types::message::MessageContentValue::Text(text) = &msg.content {
                        if !text.is_empty() {
                            blocks.push(serde_json::json!({"type": "text", "text": text}));
                        }
                    }
                    for tc in tool_calls {
                        let args: serde_json::Value = serde_json::from_str(&tc.function.arguments).unwrap_or(serde_json::Value::Object(serde_json::Map::new()));
                        blocks.push(serde_json::json!({
                            "type": "tool_use",
                            "id": tc.id,
                            "name": tc.function.name,
                            "input": args,
                        }));
                    }
                    result.push(serde_json::json!({"role": "assistant", "content": blocks}));
                    continue;
                }
            }

            result.push(serde_json::json!({"role": role, "content": content}));
        }

        if let Some(text) = current_text {
            result.push(serde_json::json!({"role": "assistant", "content": [{"type": "text", "text": text}]}));
        }

        result
    }
}

impl LlmFormatter for AnthropicFormatter {
    fn build_request(&self, request: &LlmRequest, profile: &LlmProfile) -> LlmResult<reqwest::Request> {
        let url = format!("{}/messages", profile.base_url.as_deref().unwrap_or(&self.base_url));

        let messages = self.convert_messages(&request.messages);

        let mut body = serde_json::json!({
            "model": profile.model,
            "messages": messages,
            "max_tokens": 4096,
        });

        let merged_params = crate::formatter_helpers::merge_parameters(profile, &request.parameters);
        if let Some(temp) = merged_params.get("temperature") {
            body["temperature"] = temp.clone();
        }
        if let Some(max_tokens) = merged_params.get("max_tokens") {
            body["max_tokens"] = max_tokens.clone();
        }
        if let Some(top_p) = merged_params.get("top_p") {
            body["top_p"] = top_p.clone();
        }
        if let Some(stop) = merged_params.get("stop") {
            body["stop_sequences"] = stop.clone();
        }

        if let Some(tools) = &request.tools {
            let tool_defs: Vec<serde_json::Value> = tools.iter().map(|t| {
                serde_json::json!({
                    "name": t.name,
                    "description": t.description,
                    "input_schema": t.parameters,
                })
            }).collect();
            body["tools"] = serde_json::json!(tool_defs);
        }

        if request.stream == Some(true) {
            body["stream"] = serde_json::json!(true);
        }

        let mut req_builder = reqwest::Client::new()
            .request(Method::POST, &url)
            .header("Content-Type", "application/json")
            .header("anthropic-version", "2023-06-01")
            .json(&body);

        if let Some(api_key) = &profile.api_key {
            req_builder = req_builder.header("x-api-key", api_key);
        }

        if let Some(headers) = &profile.headers {
            for (key, value) in headers.iter() {
                req_builder = req_builder.header(key, value.as_str().unwrap_or(""));
            }
        }

        req_builder.build().map_err(crate::error::LlmError::HttpError)
    }

    fn parse_response(&self, body: &str) -> LlmResult<LlmResponseType> {
        let json: serde_json::Value = serde_json::from_str(body)?;

        let id = json.get("id").and_then(|v| v.as_str()).map(String::from);
        let model = json.get("model").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let stop_reason = json.get("stop_reason").and_then(|v| v.as_str()).map(String::from);

        let content_arr = json.get("content").and_then(|v| v.as_array());
        let mut text_content = String::new();
        let mut tool_calls = Vec::new();

        if let Some(blocks) = content_arr {
            for block in blocks {
                let block_type = block.get("type").and_then(|v| v.as_str());
                match block_type {
                    Some("text") => {
                        if let Some(text) = block.get("text").and_then(|v| v.as_str()) {
                            text_content.push_str(text);
                        }
                    }
                    Some("tool_use") => {
                        let tc_id = block.get("id").and_then(|v| v.as_str()).unwrap_or("").to_string();
                        let name = block.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string();
                        let input = block.get("input").cloned().unwrap_or(serde_json::Value::Object(serde_json::Map::new()));
                        let arguments = serde_json::to_string(&input).unwrap_or_default();
                        tool_calls.push(wf_types::message::LlmToolCall {
                            id: tc_id,
                            r#type: "function".to_string(),
                            function: wf_types::message::LlmFunctionCall { name, arguments },
                        });
                    }
                    _ => {}
                }
            }
        }

        let usage = json.get("usage").map(|u| {
            let input = u.get("input_tokens").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
            let output = u.get("output_tokens").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
            wf_types::llm::TokenUsageStats {
                prompt_tokens: input,
                completion_tokens: output,
                total_tokens: input + output,
                reasoning_tokens: None,
                prompt_tokens_cost: None,
                completion_tokens_cost: None,
                total_cost: None,
            }
        });

        let message = wf_types::message::Message {
            id: wf_types::Id::new(),
            role: wf_types::message::MessageRole::Assistant,
            content: wf_types::message::MessageContentValue::Text(text_content.clone()),
            timestamp: wf_common::time::now(),
            tool_call_id: None,
            tool_name: None,
            tool_calls: if tool_calls.is_empty() { None } else { Some(tool_calls.clone()) },
            thinking: None,
            metadata: None,
        };

        let reasoning_token_count = json.get("usage")
            .and_then(|u| u.get("completion_tokens_details"))
            .and_then(|d| d.get("reasoning_tokens"))
            .and_then(|r| r.as_u64());

        let mut metadata = std::collections::HashMap::new();
        if let Some(t) = json.get("type").and_then(|v| v.as_str()) {
            metadata.insert("type".to_string(), serde_json::json!(t));
        }
        if let Some(sr) = &stop_reason {
            metadata.insert("stop_reason".to_string(), serde_json::json!(sr));
        }
        let metadata = if metadata.is_empty() { None } else { Some(metadata) };

        Ok(LlmResponseType {
            id,
            model,
            content: if text_content.is_empty() { None } else { Some(text_content) },
            message,
            tool_calls: if tool_calls.is_empty() { None } else { Some(tool_calls) },
            usage,
            finish_reason: stop_reason,
            duration: 0,
            reasoning_content: None,
            reasoning_tokens: reasoning_token_count.map(|r| r as u32),
            metadata,
            stream_stats: None,
            warnings: None,
        })
    }

    fn parse_stream_chunk(&self, data: &str) -> LlmResult<Option<MessageStreamEvent>> {
        if data.is_empty() {
            return Ok(None);
        }

        let json: serde_json::Value = serde_json::from_str(data)?;
        let event_type = json.get("type").and_then(|v| v.as_str());

        match event_type {
            Some("message_start") => Ok(None),
            Some("content_block_start") => {
                let block = json.get("content_block");
                if let Some(b) = block {
                    if b.get("type").and_then(|v| v.as_str()) == Some("tool_use") {
                        return Ok(None);
                    }
                }
                Ok(None)
            }
            Some("content_block_delta") => {
                let delta = json.get("delta");
                if let Some(d) = delta {
                    match d.get("type").and_then(|v| v.as_str()) {
                        Some("text_delta") => {
                            if let Some(text) = d.get("text").and_then(|v| v.as_str()) {
                                return Ok(Some(MessageStreamEvent::Text(
                                    wf_types::llm::MessageStreamText { text: text.to_string() }
                                )));
                            }
                        }
                        Some("thinking_delta") => {
                            if let Some(reasoning) = d.get("thinking").and_then(|v| v.as_str()) {
                                return Ok(Some(MessageStreamEvent::ReasoningText(
                                    wf_types::llm::MessageStreamReasoning { reasoning: reasoning.to_string() }
                                )));
                            }
                        }
                        _ => {}
                    }
                }
                Ok(None)
            }
            Some("message_delta") => {
                let delta = json.get("delta");
                if let Some(d) = delta {
                    if let Some(stop_reason) = d.get("stop_reason").and_then(|v| v.as_str()) {
                        let is_done = matches!(stop_reason, "end_turn" | "max_tokens" | "stop_sequence" | "tool_use");
                        if is_done {
                            return Ok(Some(MessageStreamEvent::End(wf_types::llm::MessageStreamEnd {})));
                        }
                    }
                }
                Ok(None)
            }
            Some("message_stop") => Ok(Some(MessageStreamEvent::End(wf_types::llm::MessageStreamEnd {}))),
            _ => Ok(None),
        }
    }

    fn convert_tools(&self, tools: &[Tool]) -> LlmResult<Vec<serde_json::Value>> {
        let tool_defs: Vec<serde_json::Value> = tools.iter().map(|t| {
            serde_json::json!({
                "name": t.name,
                "description": t.description,
                "input_schema": t.parameters,
            })
        }).collect();
        Ok(tool_defs)
    }

    fn parse_tool_calls(&self, result: &LlmResponseType) -> Vec<wf_types::message::LlmToolCall> {
        result.tool_calls.clone().unwrap_or_default()
    }
}
